cirq/sim/simulator.py

```diff
@@ -173,6 +173,103 @@ def compute_amplitudes_sweep(
         raise NotImplementedError()
 
 
+class SimulatesExpectationValues(metaclass=abc.ABCMeta):
+    """Simulator that computes exact expectation values of observables.
+
+    Given a circuit and an observable map, computes exact (to float precision)
+    expectation values for each observable at the end of the circuit.
+
+    Implementors of this interface should implement the
+    simulate_expectation_values_sweep method.
+    """
+
+    def simulate_expectation_values(
+        self,
+        program: 'cirq.Circuit',
+        observables: Dict[str, 'cirq.PauliSum'],
```

Comment:
huh, I didn't catch this in the design doc - why this vs just a list and the same ordering in the returned value which would be a list?

Reply:
I updated the doc after some discussion with Matt - in short, this is intended to make finding a result for a given observable less confusing, particularly in the case of `simulate_expectation_values_sweep`.

Reply:
But now the user has to come up with a naming scheme for a potentially large list of PauliSums. Why does the sweep version matter? Because it would be a list of lists?

Reply:
That was the point of concern, yes. Pinging @mpharrigan for his take on this.

Reply:
I raised a concern about list-of-lists because it's not clear which should be nested first and reading the code or type signature wouldn't help very much. The issue--of course--is that sweeps and the collections of observables both serve the same role: "Hey take this circuit [ansatz] and monkey with it". With sweeps, you can explicitly outer-product two together. Here, it's implicit. Lots of words to say I'm not the biggest fan of list-of-lists.

That being said, it would be a huge nightmare to have to come up with keys for this. On the design doc comment, there was some initial discussion about having the return value be a pandas dataframe where the sweep values and paulisum observables could coexist as columns. I don't know if there's much appetite for that. It might behoove us to think about user stories for doing a sweep of parameters on many observables. When I was making the examples for the nisq-observable-estimation routines a dataframe felt very natural but I'm a big pandas person

Reply:
One note: I had suggested `pandas.DataFrame` for the Sampler API, since other Sampler methods also return `pandas`. It could work for simulator as well, but there are (currently) no simulator methods using `pandas`.

Rough list of options (likely incomplete):
1. Keep dict -> list-of-dicts.
    a. Clarifies nesting, but adds the difficulty of key creation. (looks like `results[1]['ZXZ']`)
1. Use dict -> `pandas.DataFrame`.
     a. Adds pandas convenience methods, but contrasts with simulator methods. (looks like...?)
1. Revert to list -> list-of-lists.
     a. Easier to create input, less clarity in output. (looks like `results[1][3]`)
1. Use type aliases, e.g. `ExpectationValues = List[float]`
     a. Clarifies function signature, but not return-value usage (would still look like `result[1][3]`)
1. Use custom return types, e.g.

    ```
    @dataclass
    class ExpectationValues:
        evs: List[float]
    ```
    a. Clarifies return-value usage, but may add clutter (looks like `result[1].evs[3]`)

Thoughts? If the type signature is the main concern, option 4 is the cheapest solution.

Reply:
I'd say (3) or (2). The others add a lot of bookkeeping without really helping all that much

Reply:
SGTM. To avoid the key-generation nightmare, I'll go ahead with (3). Opened #3708 to discuss potentially shifting simulators to `pandas`.

Reply:
I'm comfortable with (3) as it's simple - but I understand the ambiguity that it introduces, and the mental bookkeeping the user has to do - they'll have to remember that sweeps are the outer index. 

I like the idea of (2) - adds more power, more structure and consistency. 

My OO java background would go for (5) to avoid [primitive obsession](https://dzone.com/articles/code-quality-fighting-primitive-obsession-code-sme-1) - it would force the user to write something like: `result.sweeps[0].evs[1]` but I'm not sure that that's "pythonic" :)